src/Simulation/CsharpGeneration/SimulationCode.fs

```diff
@@ -1412,11 +1352,81 @@ module SimulationCode =
             |> List.map buildOne
             |> List.choose id
 
-        ``#line hidden`` <|
+        let testHandles : MemberDeclarationSyntax list = 
+            let testTargets = function 
+                | QsCallable c -> 
+                    c.Attributes 
+                    |> SymbolResolution.TryFindTestTargets 
+                    |> Seq.map (fun target -> (c.FullName, c.SourceFile, fst c.Location.Offset), target) 
+                    |> Seq.toList
+                | _ -> []
+            let unitTests = // contains the full name of the callable which to test as well as the name of the target simulator
+                if globalContext.unitTests.Any() then 
+                    localElements 
+                    |> List.collect testTargets 
+                    |> List.filter (snd >> String.IsNullOrWhiteSpace >> not)
+                    |> List.sortBy (fun ((name,_,_), target) -> (name, target))
+                else []
+            let assemblyName = if globalContext.AssemblyName = null then null else globalContext.AssemblyName.Replace(".", "")
+            [
+                // For each unit test, we create something like the following code:
+                //
+                // namespace AssemblyName // we omit this nested namespace if the assembly name is not specified
+                // {
+                //   public partial class __ExecuteOnTargetName__
+                //   #line nr source
+                //   {
+                //     [Fact(DisplayName = "TestOperationName")]
+                //     public void __TestOperationNameTest__()
+                //     {
+                //         var sim = new TargetName(); 
+                //         sim.OnLog += this.Output.WriteLine;
```

Comment:
And we should check that `sim` is an instance of `SimulatorBase` before attaching to the event.